torch/autograd/functional.py

```diff
@@ -408,7 +433,65 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
     return chunks
 
 
-def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
+def _jacfwd(func, inputs, strict=False, vectorize=False):
+    if strict:
+        raise RuntimeError('torch.autograd.functional.jacobian: `strict=True` '
+                           'and `forward_ad=True` are not supported together (yet). '
+                           'Please either set `strict=False` or '
+                           '`forward_ad=False`.')
+    is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "jacobian")
+    outputs = func(*inputs)
+    is_outputs_tuple, outputs = _as_tuple(outputs,
+                                          "outputs of the user-provided function",
+                                          "jacobian")
+    if vectorize:
+        # See NOTE: [Computing jacobian with vmap and grad for multiple outputs]
+        if strict:
+            raise RuntimeError('torch.autograd.functional.jacobian: `strict=True` '
+                               'and `vectorized=True` are not supported together. '
+                               'Please either set `strict=False` or '
+                               '`vectorize=False`.')
+        input_numels = tuple(input.numel() for input in inputs)
+
+        # Step 1: Prepare tangents
+        tangents = _construct_standard_basis_for(inputs, input_numels)
+
+        # Step 2: Compute vmap over computation with dual tensors
+        def jvp(tangents):
+            with fwAD.dual_level():
+                dual_inputs = tuple(
+                    fwAD.make_dual(input, tangent.view_as(input)) for input, tangent in zip(inputs, tangents))
+                dual_outputs = _as_tuple(func(*dual_inputs))
+                jv = []
+                for dual_out in dual_outputs:
+                    primal, tangent = fwAD.unpack_dual(dual_out)
+                    if tangent is not None:
+                        jv.append(tangent)
+                    else:
+                        jv.append(torch.zeros_like(primal))
+                return tuple(jv)
+
+        outputs_before_split = _vmap(jvp)(tangents)
```

Comment:
Is this all we need to support jacobian/hessian computation? IIRC there's a restriction that the strides of the tangent need to be the same as the strides of the primal -- what is the plan to resolve that?

Reply:
Starting to look into this more. The current update is: If we define a batching rule for an op to be "stride preserving" if performing the op in a batched way produces the exact same strides in that logical tensor as performing the op in a non-batched way - 

I think the possible outcomes here are:
 - non-problem: it is hard to write batching rules that are NOT stride preserving, i.e., the only way to run into that situation is through contrived means (performing an extra copy for no reason)
 - minor problem: depending on the memory format/contiguity of the input tensor(s) certain batching rules currently do not have stride preserving batching rule implementations, but they can be tweaked to be stride preserving without incurring an extra copy.
 - major problem: there are classes of operations that don't have efficient stride preserving batching rule implementations. We must incur an extra copy in this case for that operation to maintain compatibility with forward AD. In this case, for those out-of-place ops we should differentiate between when we are performing forward-AD and perform the extra copy in the batching rule if only that is the case (and then have a warning).

On the surface it seems that "non-problem" is most likely because
- view operations are always stride preserving since they all reduce to as_strided which is stride preserving
- out-of-place operations typically produce contiguous outputs (how could they not?)

But you did mention that the batching rule for convolution could produce non-contiguous outputs... so I'm currently investigating that case to see if that could be an issue.

Reply:
>  If we define a batching rule for an op to be "stride preserving" 

We do not enforce batching rules be "stride preserving". One easier way to test this is perhaps to make some batching rule, like that of torch.sin (if it exists...) to return a tensor with different strides, and see if fw-mode AD has any problems with it.

Reply:
> major problem: there are classes of operations that don't have efficient stride preserving batching rule implementations. We must incur an extra copy in this case for that operation to maintain compatibility with forward AD. In this case, for those out-of-place ops we should differentiate between when we are performing forward-AD and perform the extra copy in the batching rule if only that is the case (and then have a warning).

Do we have to incur an extra copy? Why does the tangent need to have the same strides as the primal in the first place?

Reply:
>  Why does the tangent need to have the same strides as the primal in the first place?

It's for ops like flatten that sometimes return a view and sometimes do not to behave consistently between the primal and tangent. For example:
 - take a pair of tangent/primal strided differently, lets say one is contiguous, and the other is not
 - then do flatten on it. The result of flatten is that the contiguous produces a view, but non-contiguous produces a copy
 - so the result is that the output of flatten is somewhere between being a view and not a view

> We do not enforce batching rules be "stride preserving".

Can we though? It sounds like we don't have any batching rules that are currently known to be not stride preserving?

Reply:
>  It sounds like we don't have any batching rules that are currently known to be not stride preserving?

Whoops I take this back. After adding a per-op test (though some are skipped) it looks like there are (at least?) 28 failures out of 373 tests. 
- cases where the stride is only different when the size is 1 is filtered out, otherwise there are few more.
- only checking the case where batch_dim=0, and single tensor output is returned
- 11 of those are conv2d, other ones are mostly linalg ones
- Unrelated: I'm suspecting the batched fallback always makes the output layout contiguous (because `stack`), so any op that returns a different layout permutation would not have it be preserved

<details>
<summary>
test code
</summary>


```
    # From duplicating the test above:
    def test_vmap_strides_preservation(self, device, dtype, op):
        sample_inputs_itr = op.sample_inputs(device, dtype, requires_grad=False)
        for sample_input in sample_inputs_itr:
            arg_values = [sample_input.input] + list(sample_input.args)
            kwarg_values = sample_input.kwargs
            try:
                for loop_out, batched_out, single_out, vmap_dims in get_fallback_and_vmap_exhaustive(op.op, arg_values, kwarg_values):
                    if isinstance(batched_out, list) or isinstance(batched_out, tuple):
                        # Need for chunk, tensor_split etc.
                        pass
                    else:
                        filtered_batched_strides = tuple(i for n, i in enumerate(batched_out.stride()) if batched_out.size(n) != 1)
                        filtered_single_strides = tuple(i for n, i in enumerate(single_out.stride()) if single_out.size(n) != 1)

                        if len(vmap_dims) == 1:
                            # Only check this simple case for now bc we can assume first dim is batch dim
                            self.assertEqual(filtered_batched_strides[1:], filtered_single_strides)
            except Exception as e:
                # Checking if we're throwing an error because of dynamic shapes.
                if "dynamic" in e.args[0]:
                    continue
                raise e
```
</details>

```
tv_batch_rule
batched_out strides:  (1, 3)  single_out strides:  (1,)
batched_out sizes:  torch.Size([3, 5])  single_out sizes:  torch.Size([5])
F
======================================================================
FAIL: test_vmap_strides_preservation_matmul_cpu_float32 (__main__.TestVmapOperatorsOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/local/pytorch2/torch/testing/_internal/common_device_type.py", line 368, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/local/pytorch2/torch/testing/_internal/common_device_type.py", line 734, in test_wrapper
    return test(*args, **kwargs)
  File "test/test_vmap.py", line 3041, in test_vmap_strides_preservation
    raise e
  File "test/test_vmap.py", line 3036, in test_vmap_strides_preservation
    self.assertEqual(filtered_batched_strides[1:], filtered_single_strides)
  File "/home/local/pytorch2/torch/testing/_internal/common_utils.py", line 1943, in assertEqual
    self.assertEqual(x_, y_, atol=atol, rtol=rtol, msg=msg,
  File "/home/local/pytorch2/torch/testing/_internal/common_utils.py", line 1955, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
AssertionError: False is not true : Scalars failed to compare as equal! 3 != 1

----------------------------------------------------------------------
```

Reply:
If batch dim is not 0, then... most batching rules will move the batch dim to the front because it's easier to write a batching rule that way. That will probably cause the batching rule to be "not stride preserving"